include/trieste/logging.h

```diff
@@ -63,11 +65,19 @@ namespace trieste::logging
       // Represents same as Info and debug messages should also be
       Debug = 6,
       // Represents same as Debug and trace messages should also be printed
-      Trace = 7
+      Trace = 7,
+      // Represents an uninitialized logging level.
+      Uninitialized = 8,
     };
 
-    // Used to set which level of message should be reported.
-    inline LogLevel report_level{LogLevel::Output};
+    // Used to set which default level of message should be reported.
+    inline LogLevel default_report_level{LogLevel::Uninitialized};
+
+    static detail::LogLevel& thread_local_report_level()
+    {
+      static thread_local detail::LogLevel report_level = default_report_level == LogLevel::Uninitialized ? LogLevel::Output : default_report_level;
```

Comment:
So this is a non-trivial initialiser, this will result in a branch on every access to the `report_level` to check if it has been initialised, because C++ hates you. This puts us on two branch for every log message that is not printed.  If we detect this case in `start`, then for the not printed messages, we can get down to one branch except for the first message.